util/topgen/resets.py

```diff
@@ -0,0 +1,174 @@
+            self.clock = clocks.get_clock_by_name(raw['clk'])
+        else:
+            self.clock = None
+
+        self.sw = bool(raw.get('sw', 0))
+
+    def _asdict(self) -> Dict[str, object]:
+        ret = {
+            'name': self.name,
+            'gen': self.gen,
+            'type': self.rst_type,
+            'domains': self.domains,
+            'shadowed': self.shadowed,
+            'parent': self.parent,
+            'clock': self.clock,
+            'sw': self.sw
+        }
+
+        return ret
+
+
+class Resets:
+    '''Resets for the chip'''
+    def __init__(self, raw: Dict[str, object], clocks: Clocks):
+        self.hier_paths = {}
+        assert isinstance(raw['hier_paths'], dict)
+        for rst_src, path in raw['hier_paths'].items():
+            self.hier_paths[str(rst_src)] = str(path)
+
+        assert isinstance(raw['nodes'], list)
+
+        self.nodes = {}
+        for node in raw['nodes']:
+            assert isinstance(node, dict)
+            reset = ResetItem(node, clocks)
+            self.nodes[reset.name] = reset
+
+    def _asdict(self) -> Dict[str, object]:
+        ret = {
+            'hier_paths': self.hier_paths,
+        }
+
+        for name, node in self.nodes.items():
+            ret[name] = node._asdict()
+
+        return ret
+
+    def _get_reset_by_name(self, name: str) -> ResetItem:
+
+        ret = self.nodes.get(name, None)
+        if ret:
+            return ret
+        else:
+            raise ValueError('{name} is not a defined reset')
+
+    def set_reset_shadowed(self, name: str):
+        '''Mark particular reset as requiring shadow'''
+
+        reset = self._get_reset_by_name(name)
+        reset.shadowed = True
+
+    def get_reset_domains(self, name: str):
+        '''Get available domains for a reset'''
+
+        reset = self._get_reset_by_name(name)
+
+        # Other reset types of hardwired domains
+        return reset.domains
+
+    def get_clocks(self) -> list:
+        '''Get associated clocks'''
+
+        clocks = {}
+        for reset in self.nodes.values():
+            if reset.rst_type != 'ext':
+                clocks[reset.clock.name] = 1
+
+        return clocks.keys()
+
+    def get_generated_resets(self) -> Dict[str, object]:
+        '''Get generated resets and return dict with
+           with related clock
+        '''
+
+        ret = []
+        for reset in self.nodes.values():
+            if reset.gen:
+                entry = {}
+                entry['name'] = reset.name
+                entry['clk'] = reset.clock.name
+                entry['parent'] = reset.parent
+                entry['sw'] = reset.sw
+                ret.append(entry)
+
+        return ret
+
+    def _get_top_resets_obj(self) -> list:
+        '''Get resets pushed to the top level'''
+
+        ret = []
+        for reset in self.nodes.values():
+            if reset.rst_type == 'top':
+                ret.append(reset)
+
+        return ret
```

Comment:
This filter, and the ones below, can be written a bit more simply as list comprehensions:
```python3
    def _get_top_resets_obj(self) -> List[ResetItem]:
        '''Get resets pushed to the top level'''
        return [reset
                for reset in self.nodes.values()
                if reset.rst_type == 'top']
```

Reply:
haha yeah i actually had that at first and converted to this.
I honestly don't know how i feel about list comprehension..sometimes i feel like it's great, other times i can't make out what I was trying to type.

let me switch back.

Reply:
Happy either way, but maybe it's best to be consistent (so `get_top_resets` and `get_sw_resets` should probably match)